File: backend/app/schemas/audit.py
"""Response schemas for the unified audit timeline API."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class AuditEntryResponse(BaseModel):
    id: int
    entity_type: str
    entity_id: int
    action: str
    user_id: Optional[int] = None
    user_name: Optional[str] = None
    timestamp: datetime
    old_values: Optional[str] = None
    new_values: Optional[str] = None
    correlation_id: Optional[str] = None
    log_level: str

    class Config:
        from_attributes = True


class AuditVerifyResponse(BaseModel):
    valid: bool
    checked: int
    first_broken_id: Optional[int] = None
    # Populated only when GET /audit/verify is called with ?correlation_id=.
    correlation_entries: Optional[int] = None
    correlation_ok: Optional[bool] = None

File: frontend/src/api/audit.ts
import client from './client'

export interface AuditEntry {
  id: number
  entity_type: string
  entity_id: number
  action: string
  user_id: number | null
  user_name: string | null
  timestamp: string
  old_values: string | null
  new_values: string | null
  correlation_id: string | null
  log_level: string
}
export interface AuditVerify {
  valid: boolean
  checked: number
  first_broken_id: number | null
  // Only populated when verify() is called with a correlation_id.
  correlation_entries?: number | null
  correlation_ok?: boolean | null
}

export const auditApi = {
  list: (params: { correlation_id?: string; entity_type?: string; limit?: number; offset?: number }) =>
    client.get<AuditEntry[]>('/v1/audit', { params }).then((r) => r.data),
  verify: (params?: { correlation_id?: string }) =>
    client.get<AuditVerify>('/v1/audit/verify', { params }).then((r) => r.data),
  downloadCsv: async (params: { correlation_id?: string }) => {
    const res = await client.get('/v1/audit/export', { params, responseType: 'blob' })
    const url = URL.createObjectURL(res.data as Blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `audit_${params.correlation_id ?? 'export'}.csv`
    a.click()
    URL.revokeObjectURL(url)
  },
}
